A complete re-write of the ABNF parser.

ldthomas opened the pull request:
This is a complete rewrite of the ABNF parser in `abnf.ts`. It uses a fixed, pre-generated ABNF grammar object instead of re-generating the object each time a message is parsed. It runs approximately twice as fast as the original parser, is more accurate in URI validation and removes the dependencies on `uri-js` and `valid-url`.

Notes:
- `packages/siwe/package.json` - removed `uri-js` and `valid-url` dependencies.
  - note that a global search indicated that `uri-js` was never used anyway
  - suggest bumping the version to 2.4.0 but versioning is your prerogative
- `packages/siwe-parser/package.json` - removed `uri-js` and `valid-url` dependencies. Added a script for re-generating the fixed ABNF grammar object, `packages/siwe-parser/siwe-grammar.js`, from the ABNF grammar, `packages/siwe-parser/siwe-abnf.txt`. `npm run apg` needs to be run each time a change is made to the ABNF grammar.
  - suggest bumping the version to 2.2.0 but versioning is your prerogative
- even after deleting the dependence on `uri-js` and `valid-url`, the `package-lock.json` files still contain dozens of references to them. Deleting the `package-lock.json` files and re-generating them from scratch cleans them up considerably. However, I'm aware there may be unintended consequences from that maneuver. Just a suggestion if you want to clean them up.
- `packages/siwe/client.ts` - contains a workaround for 
> import { ParsedMessage, parseIntegerNumber } from '@spruceid/siwe-parser';
- as mentioned in issue #177, in my environment a build of the cloned version of this repository fails to find '@spruceid/siwe-parser'. I've not found a fix for this problem yet.
- a large number of unit tests have been added for testing the accuracy of parsing both the message URI and the resource URIs

sbihel commented on packages/siwe-parser/lib/abnf.ts:
```diff
@@ -1,366 +1,162 @@
+      },
+    };
+    // uncomment for tracing
+    // if (doTrace) {
```
Is there a way to automatically enable this when needed without having to uncomment code?

ldthomas replied:
Yes. I can make that happen. It will come in handy if you find yourself scratching your head over a message that you think is good but fails. You can usually spot the problem quickly in the trace. The trace will show up in a new folder `output/trace.html`.

ldthomas replied:
I just pushed changes to ldthomas/siwe that added an optional `doTrace` argument to the `ParsedMessage` constructor. If `doTrace` is `true` the trace of the parse tree will be written to `output/siwe-parser-trace.html`. Consequently, I've added `**/output*` to both `.gitignore` and `.npmignore`.

sbihel replied:
I didn't realise it was going to write to a file and make use of node dependencies -- isn't there a way to just output logs? It seems odd to me that a library would output files

ldthomas replied:
I used the trace extensively in my development and testing. However, the reason it was commented out in the first place was that I assumed that you wouldn’t want node dependencies in the production code. But if you want tracing as an option, node dependencies is the price to pay. 

For now I’m commenting the tracing code out again. However, I have a model for printing the trace to the console in real time in `apg-lite`. I’m not sure how much work it will be but I’ll take a look at moving that technique as an option into `apg-js`. I should have it fixed or will have given up on it by mid next week..

sbihel commented on packages/siwe-parser/lib/t-ipv4-ipv6.test.ts:
```diff
@@ -0,0 +1,145 @@
+};
+let result;
+
+describe("test IPv4 addresses", () => {
```
If you could add those tests to the JSON tests (which can then be used by other libraries in other languages) that would be great

ldthomas replied:
Uumm. Not so sure about this one. That looks like a ton of work creating JSON objects with all of that test data in it. But I can see out it might be helpful to your overall project beyond just this `ParsedMessage` object. We can discuss this further.

ldthomas replied:
Yes, I will try to put all of the new tests into JSON objects so that they can be run similarly to `parsers.test.ts` or maybe fully integrated into that file. But I'd like to close this conversation for now and present the modified unit tests as a separate PR later.

sbihel commented on packages/siwe/lib/client.ts:
```diff
@@ -426,21 +426,10 @@ export class SiweMessage {
-        this.address
-      );
-    }
+    // this test is done in siwe-parser, callback function "address"
```
If all the checks done in this function are done as part of the parsing, then I suggest no to modify this function but instead move its call in the constructor to only the case where it's initialised from an object.

ldthomas replied:
If I understand this correctly you still want to validate the EIP-55 address and the URI in the constructor block for the message object. The EIP-55 address will be simple enough, but I'll need to build a special validator that validates only the URI and not an entire message. The ABNF grammar object, `siwe-grammar.ts` and the existing call back functions give me everything I need for that, but it will require having an option in the `ParsedMessage` constructor to parse only the URI. How about the resource URIs? Will they need to be validated as well?

ldthomas replied:
I need to write a new `isUri()` function to use in the object block of the constructor. I just need to assemble it from existing pieces which won't be difficult. And, of course, do the unit testing. I'll need a few more days on this.

ldthomas replied:
The latest commit should resolve this conversation. An `isUri()` function has been added, unit tested and the message block of the `SiweMessage` constructor updated to test the EIP-55 address, the URI and all of the resource URIs if any are present. However, see the additional question in the following conversation.

sbihel replied:
You've extracted the URI and EIP55 validation out of the validation function, but remains:
- domain
- version
- nonce
- dates

These can (or already are) validated in the parser, so can't the validate function be left untouched and only called when constructing the message from an object?

ldthomas replied:
Sorry, I missed the proper flow of the validation in the message object case. In particular I missed the `validateMessage()` function. All I really needed to do was replace the valid-url isUri() function with the new APG isUri() function there. All other validations will then remain the same. I can easily make that change. However, I notice there is no validation of the resources URIs. Do you have any need for that? It is easily added.

Here’s a better suggestion. Remove the `validateMessage()` function altogether and replace it with
> // this.validateMessage();
> new ParsedMessage(this.prepareMessage());

I’ve tried it and it passes all of my objects unit tests. This will also validate all of the resources URIs. (NOTE: that perpareMessage() needs a fix for the “empty statement” case that #30 pointed out was allowed in the spec.)

ldthomas replied:
BTW all of the `client.test.ts` unit tests pass with this fix also. But I have a question. When I run jest on `client.test.ts` it prints out several screens of red error text and a lot of warnings that about `validate() has been deprecated,` But then at the end it shows that all tests have passed. What's a person to make of this? Does something need to be fixed here?

sbihel replied:
> it prints out several screens of red error text and a lot of warnings

This PR will require a new major version and the deprecated function will be removed then, so that will remove those warnings.

> Here’s a better suggestion. Remove the validateMessage() function altogether and replace it with `new ParsedMessage(this.prepareMessage());`

I was worried we would lose the information provided by `SiweErrorType` (e.g. if a nonce isn't long enough it would clearly state so), and we do a bit but it still points to the field (e.g. now it says something like "line 9: invalid nonce") so I'm ok with this change. But could you remove the variants from `SiweErrorType` that aren't used anymore?

Cali93 replied:
Related, EIP55 validation should be optional https://github.com/spruceid/siwe/issues/53#issuecomment-2227835751

sbihel commented on packages/siwe/lib/client.ts:
```diff
@@ -1,10 +1,8 @@
-  parseIntegerNumber,
-} from '@spruceid/siwe-parser';
-import * as uri from 'valid-url';
+// import { ParsedMessage, parseIntegerNumber } from '@spruceid/siwe-parser';
```
This is a monorepo managed with `lerna`. If you run `npm run build` or `npm run test` it will build/test all the packages in the workspace.

ldthomas replied:
Got it. It worked. Thanks.

ldthomas replied:
Running `npm run build` in the root directory fixed the
> import { ParsedMessage, parseIntegerNumber } from '@spruceid/siwe-parser';

problem. However, I have now added
> export const isUri = (uri: string, doTrace = false) => {}

to the same file, `packages/siwe-parser/lib/abnf.ts`, but 
> import { isUri } from '@spruceid/siwe-parser';

fails even after re-running `npm run build`. I suspect that there is a file somewhere that specifically adds these class and function names to be found in `@spruceid/siwe-parser` but I’ll need some help on that.

sbihel replied:
The entrypoint for the package is `lib/parsers.ts`

ldthomas replied:
Got it. Thanks.

sbihel commented on packages/siwe-parser/lib/siwe-abnf.txt:
```diff
@@ -0,0 +1,210 @@
+request-id      = *pchar
+chain-id        = 1*DIGIT
+    ; See EIP-155 for valid CHAIN_IDs.
+resources       = *( LF resource )
```
So this PR closes #30 then?

ldthomas replied:
At first glance, I think so. But I'll need a little more time to go through #30 in more detail for a definitive reply.

ldthomas replied:
Yes, issue #30 can be closed now. The `siwe-parser` matches the spec for both `statement` and `resources`. I should have read #30 more closely. The spec ABNF grammar format does indeed allow for present, empty or missing for both the `statement` and the `resources`. With the latest commit, the `siwe-parser` matches the spec for both `statement` and `resources`. There are unit tests for all cases in `t-spec.test.ts`.

sbihel reviewed (changes requested):
Thank you for the PR! 


There are some lint errors (you can run `npm run lint` at the root of the repo).

ldthomas commented on packages/siwe-parser/lib/callbacks.ts:
```diff
@@ -0,0 +1,482 @@
+        break;
+    }
+  },
+  pathAbempty: function pathAbempty(result, chars, phraseIndex, data) {
```
**Linting errors:** There are only 3 lint problems although they show up in 4 different tests. The first is a trivial warning (see toTrace discussion later) the second is a quick fix of a forgotten `break` statement. However, the file `siwe-parser/lib/siwe-grammar.js` is generated by `apg-js`. So I am in the process of updating `apg-js` (new version 4.4.0) to generate a typescript version of the grammar object file. (This is my first time using typescript, so I'm winging it a bit here.) So pushing any fixes to the fork will have to wait for the new release of `apg-js`. Shouldn't take long.

**packages/siwe-parser/lib/callbacks.ts:** No. The additional callbacks are primarily for validating the URI. They are required for identifying the URI components. Except for obvious errors, no real effort is made to pin down specific errors. For example, a message with `URL: sample.com\n` would fail but you would not get an error message telling you that you had used `URL` instead of `URI`. That's a lot of work and does require a lot of "error productions" with corresponding callbacks.

ldthomas replied:
Hi @sbihel. Fixing the `apg-js` generator to generate a typescript-compatible grammar object was not difficult. However, verifying that I didn't break anything in `apg-lite` and especially `apg-js-examples` has turned out to be a little tedious. But I'm almost there. I should have these first two conversations resolved early next week.

ldthomas replied:
I just pushed changes to ldthomas/siwe that should resolve all these issues. Let me know if you still have any of these problems.

ldthomas commented:
I think I’m done here except for 

1. I’ve included a workaround in `client.ts`. For importing `isUri`. I had to point to the specific directory and file. I’ll need some advice on how to fix that.
2. I’ve not used your JSON object file methodology for the unit tests. As is mentioned previously, I’m willing to familiarize myself with that and do the (tedious) work of moving all of the test data into the proper JSON format. But I’d like to finish here for now and leave that to a later PR.

ldthomas commented:
**Trace:** After looking at it, redesigning the trace module and publishing a new release of `apg-js` is way too much work for a feature that is not necessary for the operation of `siwe` and probably wouldn’t get much use anyway. The code is still there in comments if someone really wants to use the trace feature to debug a message.

**Date times:** In the process of analyzing the validation of the message objects in `SiweMessage` I’ve noticed that there is an asymmetry between the way the messages and message objects are currently validated. The messages are validated by parsing them but the parser does not check the semantics of the date times. RFC 3339 defines the syntax of the date time but much of the semantics is simply mentioned in comments. For example, the syntax of the date-month is simply 2 integers but the semantics requires the number to be in the range 1-12. In the case of a message object the semantics are validated with the regex `ISO8601`. I’ve copied this regex to the parser `callbacks.ts` and there I use it to validate the date time semantics while parsing. I’ve added unit tests in `siwe/lib/objects.test.ts` to cover both syntax and semantic date time errors.

**validateMessage():** I’ve removed this function altogether and replaced it by parsing the “stringified” version of the message object. In this way, there is complete symmetry between the validation of a message and a message object. (Incidentally, I also fixed a small bug in the `toMessage()` function so that the case of an empty statement is handled according to spec.)

From the root `lint` and `test` both run without errors. I hope this latest commit works for you. If so then I can move on to updating the unit testing to your JSON object methodology.

ldthomas commented:
My apologies for an oversight. I’m still getting myself up to speed with the different ways a message and  message object are handled in the `SiweMessage` constructor. I missed a bracket paring and ended up parsing the message twice. Taking a more careful look at what `validateMessage()` does, the only thing is can see that it is checking that the parser doesn’t is the validation of the date-times. `isValidISO8601Date()` does a little more than just the regex. On the other hand, there are some message fields that the parser validates that `validateMessage()` does not:

-  scheme
-  statement
-  Chain ID

So in this latest commit, I’ve moved the complete date-time validation into the parser’s callback functions and fixed the parsing/validation in the message/object blocks so that there is no redundancy. My intention here now is that all fields are both syntactically and semantically correct if the parser succeeds. 

BTW `siwe/lib/utils.ts isValidISO8601Date()`, which I copied into the parser’s callback functions has a bug. Line 67
> if (inputDate) {

should read
> if (inputMatch) {

Bettyrockergit reviewed (commented):
packages/siwe-parser/lib/t-chars.test.ts

ldthomas commented:
Thanks. Missed the debugging code left in this test.

ldthomas commented:
Hi @sbihel. Anything else you need? From my end I'm done here unless there is something else required for a merge. I'm waiting on that before I begin working on converting the unit tests to your JSON file methodology.

sbihel commented:
Thank you, it looks great. In addition to the clean-up of `SiweErrorType`:

> BTW siwe/lib/utils.ts isValidISO8601Date(), which I copied into the parser’s callback functions has a bug

Would you mind fixing it as part of this PR please?

> before I begin working on converting the unit tests to your JSON file methodology

Thank you for doing that, you can just move the tests from `objects.test.ts` and `t-spec.test.ts` -- the other tests are a bit too "low-level" and won't provide much value to other implementations.

ldthomas commented:
The fix to `siwe/lib/utils.ts` is done.

Converted `siwe/lib/objects.test.ts` to JSON file methodology.

I had already converted all of the `siwe-parser/lib/t-*.test.ts` to the JSON methodology and they are all included in `unit.test.ts`. If others are only interested in the specification tests they only need to use the `valid-specification.json` file and ignore the others. But I think having all of the unit tests available is nice for a quick check that anytime a change is made it hasn’t broken anything.

I really don’t want to mess with the `SiweErrorType` enum. The `siwe-parser`, past and present, never uses them. They only seem to be used in the siwe client and since I don’t really know what is going on there I really need to stay out of it. However, looking into that I did notice a big problem in `client.test.ts`. I’m not great with `async` programming so I’m not going to attempt to fix it but have a look at this code:
> (n, test) => {
>   try {
>     new SiweMessage(test);
>   } catch (error) {
>     expect(Object.values(SiweErrorType).includes(error));
>   }
> }
It can never fail and really does no testing at all. 1) If for some reason SiweMessage() unexpectedly succeeds, the test will just fall through without announcing failure. 2) `expect(true)` and `expect(false)`, without any expect matchers, both succeed. So no matter what the value of `Object.values(SiweErrorType).includes(error)` the test succeeds. Not to mention that `SiweMessage()` doesn’t throw `SiweErrorType` anymore. My suggestion is to just replace this with an `expect().toThrow()` test. But I really don’t know what is going on with the tests that use `async` so I’m not going to touch that. Just pointing it out.

ldthomas commented:
I'm withdrawing this PR as it seems not to be a good fit. I've recently published a stand-alone siwe message parser [here on npm](https://www.npmjs.com/package/parse-siwe) and [here on GitHub](https://github.com/ldthomas/parse-siwe). You may possibly find it to be a more useful alternative.

w4ll3 commented:
Hi there, actually I'm planning to merge it, was just doing some
adjustments to it. If you don't mind keeping it open.

On Sun, Dec 15, 2024 at 6:31 PM Lowell D. Thomas ***@***.***>
wrote:

> I'm withdrawing this PR as it seems not to be a good fit. I've recently
> published a stand-alone siwe message parser here on npm
> <https://www.npmjs.com/package/parse-siwe> and here on GitHub
> <https://github.com/ldthomas/parse-siwe>. You may possibly find it to be
> a more useful alternative.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/spruceid/siwe/pull/200#issuecomment-2544076778>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACBSTQEOJL7DDOAAF3I6V2T2FXYMDAVCNFSM6AAAAABG62PPUSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKNBUGA3TMNZXHA>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

ldthomas commented:
Delighted. It's just that after 7 months of silence I assumed it was dead in the water and moved on to a separate, stand-alone project.

ldthomas commented:
Sorry if that response came across as disrespectful. Didn't mean it that way. If you need any further help don't hesitate to ask.

w4ll3 commented:
Thanks, no worries we are really glad for your contribution and actually
sorry if letting this stalled so long also bothered you.

On Sun, Dec 15, 2024, 21:31 Lowell D. Thomas ***@***.***>
wrote:

> Sorry if that response came across as disrespectful. Didn't mean it that
> way. If you need any further help don't hesitate to ask.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/spruceid/siwe/pull/200#issuecomment-2544239052>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACBSTQFS4NMOFADV5W25OND2FYNP5AVCNFSM6AAAAABG62PPUSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKNBUGIZTSMBVGI>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

w4ll3 commented:
I completely missed that the work of this branch was addressed at #213, sorry for making you reopen the PR @ldthomas. Thanks again for the contribution, closing it since it was merged.
